File: sdks/php/src/Providers/OpenAI.php
<?php

declare(strict_types=1);

namespace CyberArmor\Providers;

use CyberArmor\Client;
use CyberArmor\Policy\Decision;
use CyberArmor\Policy\PolicyViolationException;

/**
 * Wraps the openai-php/client library with CyberArmor policy enforcement and
 * audit emission.
 *
 * The underlying library must be installed separately:
 *   composer require openai-php/client
 *
 * Usage:
 *   $cyberarmor = new \CyberArmor\Client();
 *   $rawOpenAI  = \OpenAI::client(getenv('OPENAI_API_KEY'));
 *
 *   $ai = new \CyberArmor\Providers\OpenAI(
 *       client:   $cyberarmor,
 *       openai:   $rawOpenAI,
 *       tenantId: 'acme-corp'
 *   );
 *
 *   $response = $ai->chat(
 *       messages: [['role' => 'user', 'content' => 'Hello AI']],
 *       model:    'gpt-4o'
 *   );
 */
final class OpenAI
{
    /**
     * @param Client $client    The CyberArmor policy/audit client.
     * @param object $openai    An initialised openai-php/client instance.
     * @param string $tenantId  Tenant scope for policy evaluation.
     * @param string $model     Default model to use when not specified per-call.
     */
    public function __construct(
        private readonly Client $client,
        private readonly object $openai,
        private readonly string $tenantId,
        private readonly string $model = 'gpt-4o',
    ) {}

    /**
     * Check policy then forward the chat completion request to OpenAI.
     *
     * @param array<array{role: string, content: string}> $messages
     * @param string|null                                  $model    Override default model.
     * @param array<string, mixed>                         $options  Additional parameters.
     *
     * @return array<string, mixed>  Decoded OpenAI response.
     * @throws PolicyViolationException  When the policy engine denies the request.
     */
    public function chat(
        array   $messages,
        ?string $model   = null,
        array   $options = [],
    ): array {
        $resolvedModel = $model ?? $this->model;
        $prompt        = $this->extractPrompt($messages);

        $decision = $this->client->checkPolicy($prompt, $resolvedModel, 'openai', $this->tenantId);

        $resolvedMessages = $messages;
        if ($decision->decisionType === 'ALLOW_WITH_REDACTION' && $decision->redactedPrompt !== null) {
            $resolvedMessages = $this->applyRedaction($messages, $decision->redactedPrompt);
        }

        $params = array_merge($options, [
            'model'    => $resolvedModel,
            'messages' => $resolvedMessages,
        ]);

        $startTime = microtime(true);

        try {
            // openai-php/client: $openai->chat()->create($params)
            $response = $this->openai->chat()->create($params);
        } catch (\Throwable $e) {
            $this->emitAudit('completion_error', [
                'model'       => $resolvedModel,
                'prompt'      => $prompt,
                'decision'    => $decision->toArray(),
                'error'       => $e->getMessage(),
                'duration_ms' => $this->elapsedMs($startTime),
            ]);
            throw $e;
        }

        // Convert to array if the SDK returns an object/DTO.
        $responseArray = is_array($response) ? $response : (array) $response;

        $this->emitAudit('completion_returned', [
            'model'       => $resolvedModel,
            'prompt'      => $prompt,
            'decision'    => $decision->toArray(),
            'response_id' => $responseArray['id'] ?? null,
            'usage'       => $responseArray['usage'] ?? null,
            'duration_ms' => $this->elapsedMs($startTime),
        ]);

        return $responseArray;
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    /**
     * @param array<array{role: string, content: string}> $messages
     */
    private function extractPrompt(array $messages): string
    {
        $parts = [];
        foreach ($messages as $msg) {
            if (($msg['role'] ?? '') === 'user') {
                $parts[] = $msg['content'] ?? '';
            }
        }
        return implode("\n", $parts);
    }

    /**
     * Replace the last user message with the redacted prompt.
     *
     * @param array<array{role: string, content: string}> $messages
     * @return array<array{role: string, content: string}>
     */
    private function applyRedaction(array $messages, string $redactedPrompt): array
    {
        $lastUserIdx = null;
        foreach (array_reverse(array_keys($messages)) as $idx) {
            if (($messages[$idx]['role'] ?? '') === 'user') {
                $lastUserIdx = $idx;
                break;
            }
        }

        if ($lastUserIdx !== null) {
            $messages[$lastUserIdx]['content'] = $redactedPrompt;
        }

        return $messages;
    }

    /**
     * @param array<string, mixed> $payload
     */
    private function emitAudit(string $eventType, array $payload): void
    {
        $event = [
            'agent_id'    => $this->client->getAgentId(),
            'tenant_id'   => $this->tenantId,
            'event_type'  => $eventType,
            'timestamp'   => (new \DateTimeImmutable('now', new \DateTimeZone('UTC')))->format('Y-m-d\TH:i:s.v\Z'),
            'sdk_version' => '2.0.0',
            'sdk_lang'    => 'php',
            'payload'     => $payload,
        ];

        $this->client->emitAudit($event);
    }

    private function elapsedMs(float $startTime): int
    {
        return (int) round((microtime(true) - $startTime) * 1000);
    }
}

File: sdks/rust/src/providers/openai.rs
// cyberarmor-sdk/src/providers/openai.rs
//
// CyberArmor-wrapped OpenAI chat completions provider.

use std::sync::Arc;
use std::time::Instant;

use serde_json::{json, Value};
use tracing::error;

use crate::client::CyberArmorClient;
use crate::error::CyberArmorError;
use crate::policy::DecisionType;

const OPENAI_API_BASE: &str = "https://api.openai.com";
const OPENAI_DEFAULT_MODEL: &str = "gpt-4o";

/// A CyberArmor-aware wrapper around the OpenAI Chat Completions API.
///
/// All requests are checked against the CyberArmor policy engine before being
/// forwarded to OpenAI.  Audit events are emitted on success and failure.
///
/// # Example
///
/// ```no_run
/// use std::sync::Arc;
/// use cyberarmor_sdk::{CyberArmorClient, providers::openai::CyberArmorOpenAI};
/// use serde_json::json;
///
/// #[tokio::main]
/// async fn main() -> Result<(), Box<dyn std::error::Error>> {
///     let cyberarmor = Arc::new(CyberArmorClient::from_env());
///     let openai = CyberArmorOpenAI::new(
///         cyberarmor,
///         std::env::var("OPENAI_API_KEY").unwrap(),
///         "acme-corp",
///         None,
///     );
///
///     let response = openai.chat_completions(json!({
///         "model":    "gpt-4o",
///         "messages": [{ "role": "user", "content": "Hello, AI!" }]
///     })).await?;
///
///     println!("{}", response);
///     Ok(())
/// }
/// ```
#[derive(Clone, Debug)]
pub struct CyberArmorOpenAI {
    /// The underlying CyberArmor client for policy checks and audit emission.
    client: Arc<CyberArmorClient>,

    /// OpenAI API key.
    api_key: String,

    /// OpenAI API base URL (allows overriding for proxies / Azure OpenAI).
    base_url: String,

    /// Default tenant ID for policy evaluation.
    tenant_id: String,

    /// Default model to use when not specified in the request body.
    default_model: String,
}

impl CyberArmorOpenAI {
    /// Create a new `CyberArmorOpenAI` provider.
    ///
    /// # Arguments
    ///
    /// * `client`        — shared CyberArmor client
    /// * `api_key`       — OpenAI API key
    /// * `tenant_id`     — tenant identifier for policy evaluation
    /// * `base_url`      — optional override for the OpenAI base URL
    pub fn new(
        client: Arc<CyberArmorClient>,
        api_key: impl Into<String>,
        tenant_id: impl Into<String>,
        base_url: Option<String>,
    ) -> Self {
        Self {
            client,
            api_key: api_key.into(),
            base_url: base_url.unwrap_or_else(|| OPENAI_API_BASE.to_string()),
            tenant_id: tenant_id.into(),
            default_model: OPENAI_DEFAULT_MODEL.to_string(),
        }
    }

    /// Override the default model.
    pub fn with_default_model(mut self, model: impl Into<String>) -> Self {
        self.default_model = model.into();
        self
    }

    // -------------------------------------------------------------------------
    // Public API
    // -------------------------------------------------------------------------

    /// Check policy then forward a chat completion request to OpenAI.
    ///
    /// `request` must be a JSON object matching the OpenAI Chat Completions API
    /// request schema.  At minimum it must contain a `"messages"` array.
    ///
    /// The `"model"` field will be set to the default model if not present.
    ///
    /// # Errors
    ///
    /// - [`CyberArmorError::PolicyViolation`] if the request is denied in
    ///   enforce mode.
    /// - [`CyberArmorError::HttpError`] if the OpenAI HTTP call fails.
    /// - [`CyberArmorError::SerdeError`] on JSON decode failure.
    pub async fn chat_completions(
        &self,
        request: Value,
    ) -> Result<Value, CyberArmorError> {
        let mut req_obj = match request {
            Value::Object(m) => m,
            other => {
                return Err(CyberArmorError::ConfigError(format!(
                    "chat_completions expects a JSON object, got: {other}"
                )));
            }
        };

        // Ensure model is set.
        if !req_obj.contains_key("model") {
            req_obj.insert("model".to_string(), Value::String(self.default_model.clone()));
        }

        let model    = req_obj["model"].as_str().unwrap_or(&self.default_model).to_string();
        let messages = req_obj.get("messages").cloned().unwrap_or(Value::Array(vec![]));
        let prompt   = self.extract_prompt(&messages);

        let decision = self.client.check_policy(
            &prompt,
            &model,
            "openai",
            &self.tenant_id,
        ).await?;

        // Apply redaction if needed.
        if decision.decision_type == DecisionType::AllowWithRedaction {
            if let Some(ref redacted) = decision.redacted_prompt {
                req_obj.insert(
                    "messages".to_string(),
                    self.apply_redaction(&messages, redacted),
                );
            }
        }

        let start = Instant::now();
        let body  = Value::Object(req_obj);

        let url = format!(
            "{}/v1/chat/completions",
            self.base_url.trim_end_matches('/')
        );

        let result = self
            .client
            .http
            .post(&url)
            .bearer_auth(&self.api_key)
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await;

        let duration_ms = start.elapsed().as_millis() as u64;

        match result {
            Ok(resp) => {
                let status = resp.status();
                if !status.is_success() {
                    let err_body = resp.text().await.unwrap_or_default();
                    let err_str = format!("OpenAI HTTP {}: {}", status.as_u16(), err_body);

                    self.emit_audit("completion_error", json!({
                        "model":       model,
                        "prompt":      prompt,
                        "decision":    serde_json::to_value(&decision).unwrap_or_default(),
                        "error":       err_str,
                        "duration_ms": duration_ms,
                    })).await;

                    return Err(CyberArmorError::UnexpectedStatus {
                        status: status.as_u16(),
                        body: err_body,
                    });
                }

                let response: Value = resp.json().await?;

                self.emit_audit("completion_returned", json!({
                    "model":       model,
                    "prompt":      prompt,
                    "decision":    serde_json::to_value(&decision).unwrap_or_default(),
                    "response_id": response.get("id"),
                    "usage":       response.get("usage"),
                    "duration_ms": duration_ms,
                })).await;

                Ok(response)
            }
            Err(e) => {
                self.emit_audit("completion_error", json!({
                    "model":       model,
                    "prompt":      prompt,
                    "decision":    serde_json::to_value(&decision).unwrap_or_default(),
                    "error":       e.to_string(),
                    "duration_ms": duration_ms,
                })).await;

                Err(CyberArmorError::HttpError(e))
            }
        }
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    /// Extract a single string from a messages array for policy evaluation.
    fn extract_prompt(&self, messages: &Value) -> String {
        let arr = match messages.as_array() {
            Some(a) => a,
            None => return String::new(),
        };

        arr.iter()
            .filter(|m| m.get("role").and_then(Value::as_str) == Some("user"))
            .filter_map(|m| m.get("content").and_then(Value::as_str))
            .collect::<Vec<_>>()
            .join("\n")
    }

    /// Replace the last user message with the redacted prompt.
    fn apply_redaction(&self, messages: &Value, redacted_prompt: &str) -> Value {
        let mut arr = match messages.as_array() {
            Some(a) => a.clone(),
            None => return messages.clone(),
        };

        let last_user_idx = arr
            .iter()
            .enumerate()
            .rev()
            .find(|(_, m)| m.get("role").and_then(Value::as_str) == Some("user"))
            .map(|(i, _)| i);

        if let Some(idx) = last_user_idx {
            if let Some(obj) = arr[idx].as_object_mut() {
                obj.insert("content".to_string(), Value::String(redacted_prompt.to_string()));
            }
        }

        Value::Array(arr)
    }

    /// Emit an audit event; failures are logged but non-fatal.
    async fn emit_audit(&self, event_type: &str, payload: Value) {
        if let Err(e) = self.client.emit_audit_event(&self.tenant_id, event_type, payload).await {
            error!("[CyberArmor] Audit emission failed: {}", e);
        }
    }
}
